File: tests/native/capability-lease/CapabilityLeaseProof.java
import org.nostrocket.hypergolic.host.CapabilityLeaseRegistry;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.CountDownLatch;
import java.util.ArrayList;

public final class CapabilityLeaseProof {
  private static int checks = 0;
  private static void check(boolean value, String message) {
    if (!value) throw new AssertionError(message);
    checks++;
  }
  public static void main(String[] args) throws Exception {
    AtomicLong clock = new AtomicLong(1000);
    CapabilityLeaseRegistry leases = new CapabilityLeaseRegistry(clock::get);
    check(leases.register("generation-a"), "register");
    check(!leases.register("generation-a"), "duplicate registration");
    check(!leases.register("bad/path"), "invalid generation");
    check(leases.admit("foreign", 1, "snapshot") == null, "unknown generation");
    check(leases.admit("generation-a", 2, "snapshot") == null, "out-of-order sequence");
    String token = leases.admit("generation-a", 1, "immutable snapshot");
    check(token != null, "admit");
    check(!leases.isActive(token), "unclaimed token");
    check("immutable snapshot".equals(leases.take(token)), "original snapshot");
    check(leases.take(token) == null, "single take");
    check(leases.isActive(token), "claimed token");
    check(leases.finish(token), "complete once");
    check(!leases.finish(token) && !leases.isActive(token), "no reuse");
    check(leases.admit("generation-a", 1, "changed") == null, "transport replay");
    String expires = leases.admit("generation-a", 2, "expiry");
    leases.take(expires); clock.addAndGet(24_999);
    check(leases.isActive(expires), "before exact expiry");
    clock.incrementAndGet();
    check(!leases.isActive(expires) && !leases.finish(expires), "exact native expiry");
    String revoked = leases.admit("generation-a", 3, "revoked"); leases.take(revoked); leases.revoke("generation-a");
    check(!leases.sessionActive("generation-a") && !leases.isActive(revoked), "synchronous revocation");
    check(leases.take(revoked) == null && !leases.finish(revoked), "revoked result");

    CapabilityLeaseRegistry bounded = new CapabilityLeaseRegistry(clock::get);
    bounded.register("one"); bounded.register("two"); bounded.register("three");
    ArrayList<String> pending = new ArrayList<>();
    for (int i = 1; i <= 4; i++) { pending.add(bounded.admit("one", i, "x")); check(pending.get(i - 1) != null, "per-view admit"); }
    check(bounded.admit("one", 5, "overflow") == null, "per-view bound");
    for (int i = 1; i <= 4; i++) check(bounded.admit("two", i, "x") != null, "global admit");
    check(bounded.admit("three", 1, "overflow") == null, "global bound");
    bounded.take(pending.get(0)); bounded.finish(pending.get(0));
    check(bounded.admit("one", 6, "after consumed overload") != null, "sequence advances after overload");
    bounded.revoke("two");
    check(bounded.admit("three", 2, "released capacity") != null, "revocation releases capacity");
    clock.addAndGet(25_000);
    check(bounded.admit("three", 3, "expired capacity") != null, "expiry releases capacity");
    check(bounded.admit("three", 4, "🧪".repeat(CapabilityLeaseRegistry.MAX_BYTES / 4 + 1)) == null, "UTF-8 byte bound");
    check(bounded.admit("three", 5, "😀".repeat(CapabilityLeaseRegistry.MAX_BYTES / 4)) != null, "exact UTF-8 boundary");

    CapabilityLeaseRegistry concurrent = new CapabilityLeaseRegistry(clock::get);
    concurrent.register("concurrent"); String race = concurrent.admit("concurrent", 1, "one claimant");
    CountDownLatch start = new CountDownLatch(1); AtomicInteger winners = new AtomicInteger();
    ArrayList<Thread> threads = new ArrayList<>();
    for (int i = 0; i < 16; i++) {
      Thread thread = new Thread(() -> { try { start.await(); if (concurrent.take(race) != null) winners.incrementAndGet(); } catch (InterruptedException error) { throw new AssertionError(error); } });
      threads.add(thread); thread.start();
    }
    start.countDown(); for (Thread thread : threads) thread.join();
    check(winners.get() == 1, "concurrent single-use claim");
    concurrent.revoke("concurrent"); check(!concurrent.finish(race), "concurrent revoke invalidates claimant");
    CapabilityLeaseRegistry sessions = new CapabilityLeaseRegistry(clock::get);
    for (int i = 0; i < 64; i++) check(sessions.register("session-" + i), "session admit");
    check(!sessions.register("session-overflow"), "session bound");
    sessions.revoke("session-0"); check(sessions.register("fresh-generation"), "fresh generation capacity");
    System.out.println("{\"language\":\"Java\",\"checks\":" + checks + ",\"failed\":0}");
  }
}

File: tests/native/capability-lease/CapabilityLeaseProof.swift
import Foundation

private final class TestClock: @unchecked Sendable {
  private let lock = NSLock()
  private var value = ContinuousClock.now
  func now() -> ContinuousClock.Instant { lock.withLock { value } }
  func advance(_ milliseconds: Int64) { lock.withLock { value = value.advanced(by: .milliseconds(milliseconds)) } }
}
private final class Counter: @unchecked Sendable {
  private let lock = NSLock()
  private var value = 0
  func increment() { lock.withLock { value += 1 } }
  func read() -> Int { lock.withLock { value } }
}
@main
private struct CapabilityLeaseProof {
  static func main() throws {
    var checks = 0
    func check(_ value: Bool, _ message: String) throws {
      guard value else { throw NSError(domain: message, code: 1) }
      checks += 1
    }
    let clock = TestClock()
    let leases = CapabilityLeaseRegistry(now: clock.now)
    try check(leases.register("generation-a"), "register")
    try check(!leases.register("generation-a"), "duplicate registration")
    try check(!leases.register("bad/path"), "invalid generation")
    try check(leases.admit("foreign", sequence: 1, snapshot: "snapshot") == nil, "unknown generation")
    try check(leases.admit("generation-a", sequence: 2, snapshot: "snapshot") == nil, "out-of-order sequence")
    let admitted = leases.admit("generation-a", sequence: 1, snapshot: "immutable snapshot")
    try check(admitted != nil, "admit")
    let token = admitted!
    try check(!leases.isActive(token), "unclaimed token")
    try check(leases.take(token) == "immutable snapshot", "original snapshot")
    try check(leases.take(token) == nil, "single take")
    try check(leases.isActive(token), "claimed token")
    try check(leases.finish(token), "complete once")
    try check(!leases.finish(token) && !leases.isActive(token), "no reuse")
    try check(leases.admit("generation-a", sequence: 1, snapshot: "changed") == nil, "transport replay")
    let expires = leases.admit("generation-a", sequence: 2, snapshot: "expiry")!
    _ = leases.take(expires); clock.advance(24_999)
    try check(leases.isActive(expires), "before exact expiry")
    clock.advance(1)
    try check(!leases.isActive(expires) && !leases.finish(expires), "exact native expiry")
    let revoked = leases.admit("generation-a", sequence: 3, snapshot: "revoked")!
    _ = leases.take(revoked); leases.revoke("generation-a")
    try check(!leases.sessionActive("generation-a") && !leases.isActive(revoked), "synchronous revocation")
    try check(leases.take(revoked) == nil && !leases.finish(revoked), "revoked result")

    let bounded = CapabilityLeaseRegistry(now: clock.now)
    _ = bounded.register("one"); _ = bounded.register("two"); _ = bounded.register("three")
    var pending: [String] = []
    for i in 1...4 {
      let token = bounded.admit("one", sequence: UInt64(i), snapshot: "x")
      try check(token != nil, "per-view admit"); pending.append(token!)
    }
    try check(bounded.admit("one", sequence: 5, snapshot: "overflow") == nil, "per-view bound")
    for i in 1...4 { try check(bounded.admit("two", sequence: UInt64(i), snapshot: "x") != nil, "global admit") }
    try check(bounded.admit("three", sequence: 1, snapshot: "overflow") == nil, "global bound")
    _ = bounded.take(pending[0]); _ = bounded.finish(pending[0])
    try check(bounded.admit("one", sequence: 6, snapshot: "after overload") != nil, "sequence advances after overload")
    bounded.revoke("two")
    try check(bounded.admit("three", sequence: 2, snapshot: "released") != nil, "revocation releases capacity")
    clock.advance(25_000)
    try check(bounded.admit("three", sequence: 3, snapshot: "expired") != nil, "expiry releases capacity")
    try check(bounded.admit("three", sequence: 4, snapshot: String(repeating: "🧪", count: CapabilityLeaseRegistry.maxBytes / 4 + 1)) == nil, "UTF-8 byte bound")
    try check(bounded.admit("three", sequence: 5, snapshot: String(repeating: "😀", count: CapabilityLeaseRegistry.maxBytes / 4)) != nil, "exact UTF-8 boundary")

    let concurrent = CapabilityLeaseRegistry(now: clock.now)
    _ = concurrent.register("concurrent")
    let race = concurrent.admit("concurrent", sequence: 1, snapshot: "one claimant")!
    let winners = Counter()
    DispatchQueue.concurrentPerform(iterations: 16) { _ in if concurrent.take(race) != nil { winners.increment() } }
    try check(winners.read() == 1, "concurrent single-use claim")
    concurrent.revoke("concurrent")
    try check(!concurrent.finish(race), "concurrent revoke invalidates claimant")
    let sessions = CapabilityLeaseRegistry(now: clock.now)
    for i in 0..<64 { try check(sessions.register("session-\(i)"), "session admit") }
    try check(!sessions.register("session-overflow"), "session bound")
    sessions.revoke("session-0")
    try check(sessions.register("fresh-generation"), "fresh generation capacity")
    print("{\"language\":\"Swift\",\"checks\":\(checks),\"failed\":0}")
  }
}
